File: my-folder/0543-diameter-of-binary-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.diameter=0
        def len_st(node):
            if not node:
                return 0 
            left=len_st(node.left)
            right=len_st(node.right)
            self.diameter=max(self.diameter,left+right) # computes the max of lst+rst and diameter and stores in the diameter
            return max(left,right)+1 #returns the height of maxheight of lst and rst +1
        len_st(root)
        return self.diameter

        

File: my-folder/0543-diameter-of-binary-tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int diameter = 0;
    public int diameterOfBinaryTree(TreeNode root) {
        lengthOfBinaryTree(root);
        return diameter;
    }

    private int lengthOfBinaryTree(TreeNode node){
        if(node == null){
            return 0;
        }
        int leftSTLen = lengthOfBinaryTree(node.left);
        int rightSTLen = lengthOfBinaryTree(node.right); 
        diameter = Math.max(diameter, leftSTLen + rightSTLen);
        return 1 + Math.max(leftSTLen, rightSTLen);
    }
}
